docs/design/sql/08-expressions.md

```diff
@@ -0,0 +1,192 @@
+# SQL Expressions
+
+## Overview
+
+Expressions is one of the core elements of every SQL implementation.
+Expressions can participate in projections (`SELECT a + b FROM t`) and
+filters (`SELECT * FROM t WHERE a + b < 10`). Expressions may be roughly
+subdivided into 4 general categories:
+
+- Terminal expressions not referring to other expressions: column,
+  literal, parameter and type expressions.
+- Operators, like numeric arithmetic operators: `+`, `-`, etc.
+- Predicates, like comparison predicates: `<`, `>`, etc.
+- Functions, like `ABS`, `SIN`, etc.
+
+## Implementation Details
+
+We use facilities provided by Calcite to support expression parsing and
+validation. Every SQL expression goes through the following stages:
+
+- Parsing
+- Validation
+- Conversion
+- Translation
+- Evaluation
+
+### Parsing Stage
+
+The goal of the parsing stage is to obtain an AST expression
+representation from the textual expression representation for
+syntactically valid constructs and report errors for invalid ones. For
+instance, `a + b` is syntactically valid and `a +` is invalid.
+
+In Calcite, AST nodes are represented as subclasses of `SqlNode`. Column
+and type references are represented by `SqlIdentifier`. Literals are
+represented by `SqlLiteral` and its subclasses. Parameters are
+represented by `SqlDynamicParam`. Most other, if not all, expressions
+are subclasses of `SqlCall`.
+
+### Validation Stage
+
+The goal of the validation stage is to make sure that the AST
+expressions produced by the previous parsing stage are semantically
+valid. For instance, `1 + 1` is semantically valid while `1 + true` is
+not.
+
+In Calcite, the validation logic is encapsulated and orchestrated by
+implementations of `SqlValidator`. Specifically, `SqlValidatorImpl`
+extended by `HazelcastSqlValidator` is used to customize the validation
+logic to align it with our requirements.
```

Comment:
Can we define these requirements?

Reply:
clarified in d1d2929413835f7a462bc88686630dd57203b27c